File: security-adapter-in/src/main/java/com/example/security/adapter/in/rest/dto/StatsResponse.java
package com.example.security.adapter.in.rest.dto;

import com.example.security.application.query.StatsResult;
import java.time.Instant;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

public record StatsResponse(
    List<TimePoint> series, Map<String, List<TimePoint>> topGroups) {

  public static StatsResponse from(StatsResult r) {
    return new StatsResponse(
        r.series().stream().map(TimePoint::from).toList(),
        r.topGroups().entrySet().stream()
            .collect(
                Collectors.toMap(
                    Map.Entry::getKey,
                    e -> e.getValue().stream().map(TimePoint::from).toList())));
  }

  public record TimePoint(Instant timestamp, long count, double p95LatencyMs) {

    static TimePoint from(StatsResult.TimeBucket b) {
      return new TimePoint(b.timestamp(), b.count(), b.p95LatencyMs());
    }
  }
}

File: security-adapter-in/src/main/kotlin/com/example/security/adapter/in/rest/dto/StatsResponse.kt
package com.example.security.adapter.`in`.rest.dto

import com.example.security.application.query.StatsResult
import java.time.Instant

/**
 * GET /api/v1/stats 응답 — 시계열 + 그룹별 시계열.
 *
 * application layer 의 StatsResult 를 wire format 으로 평탄화한다.
 */
@JvmRecord
data class StatsResponse(
    val series: List<TimePoint>,
    val topGroups: Map<String, List<TimePoint>>,
) {

    @JvmRecord
    data class TimePoint(val timestamp: Instant, val count: Long, val p95LatencyMs: Double) {
        companion object {
            @JvmStatic
            fun from(b: StatsResult.TimeBucket): TimePoint =
                TimePoint(b.timestamp, b.count, b.p95LatencyMs)
        }
    }

    companion object {
        @JvmStatic
        fun from(r: StatsResult): StatsResponse =
            StatsResponse(
                r.series.map { TimePoint.from(it) },
                r.topGroups.mapValues { (_, list) -> list.map { TimePoint.from(it) } },
            )
    }
}
